src/main/java/coupon/global/DataSourceConfig.java

```diff
@@ -0,0 +1,52 @@
+package coupon.global;
+
+import com.zaxxer.hikari.HikariDataSource;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.jdbc.DataSourceBuilder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.DependsOn;
+import org.springframework.context.annotation.Primary;
+import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy;
+import javax.sql.DataSource;
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+public class DataSourceConfig {
+
+    @Bean
+    @ConfigurationProperties(prefix = "coupon.datasource.writer")
+    public DataSource writeDataSource() {
+        return DataSourceBuilder.create().type(HikariDataSource.class).build();
+    }
+
+    @Bean
+    @ConfigurationProperties(prefix = "coupon.datasource.reader")
+    public DataSource readDataSource() {
+        return DataSourceBuilder.create().type(HikariDataSource.class).build();
+    }
+
+    @Bean
+    @DependsOn({"writeDataSource", "readDataSource"})
+    public DataSource routeDataSource() {
+        DataSourceRouter dataSourceRouter = new DataSourceRouter();
+        DataSource writeDataSource = writeDataSource();
+        DataSource readDataSource = readDataSource();
+
+        Map<Object, Object> dataSourceMap = new HashMap<>();
+        dataSourceMap.put(DataSourceType.WRITER, writeDataSource);
+        dataSourceMap.put(DataSourceType.READER, readDataSource);
+        dataSourceRouter.setTargetDataSources(dataSourceMap);
+        dataSourceRouter.setDefaultTargetDataSource(writeDataSource);
+
+        return dataSourceRouter;
+    }
+
+    @Bean
+    @Primary
+    @DependsOn({"routeDataSource"})
+    public DataSource dataSource() {
+        return new LazyConnectionDataSourceProxy(routeDataSource());
+    }
```

Comment:
[사소한 제안] (공감가시는 부분이 있을 때만 반영해주세요~)

- 메서드 이름과 변수명이 같다면 직접 사용해도 좋을 것 같아요!
- LazyConnectionDataSourceProxy는 Proxy라고 적혀 있으니까 굳이 분리하지 않아도 좋을 것 같네요!

```suggestion
    @Bean
    @Primary
    @DependsOn({"writeDataSource", "readDataSource"})
    public DataSource dataSource() {
        Map<Object, Object> dataSources = new HashMap<>();
        dataSources.put(DataSourceType.WRITER, writeDataSource());
        dataSources.put(DataSourceType.READER, readDataSource());
        
        DataSourceRouter dataSourceRouter = new DataSourceRouter();
        dataSourceRouter.setTargetDataSources(dataSourceMap);
        dataSourceRouter.setDefaultTargetDataSource(writeDataSource);

        return new LazyConnectionDataSourceProxy(dataSourceRouter);
    }
```

Reply:
좀 더 간결한 코드네요! 두 코드의 명확한 장단점을 못 느껴서 일단 커찬의 것을 반영해 보았어요 
추후 다음 단계를 진행하며 체감을 할 수 있게 될지도...